tokio-stream/src/stream_ext/partition.rs

```diff
@@ -0,0 +1,143 @@
+#![allow(dead_code)]
+
+use crate::Stream;
+
+use std::{
+    fmt,
+    pin::Pin,
+    sync::{Arc, Mutex},
+    task::{Context, Poll, Waker},
+};
+
+/// A stream returned by the [`partition`](super::StreamExt::partition) method.
+pub enum Partition<St, F>
+where
+    St: Stream,
+{
+    /// A stream that yields items for which the predicate returns `true`.
+    Matches(Arc<Mutex<Inner<St, F>>>),
+    /// A stream that yields items for which the predicate returns `false`.
+    NonMatches(Arc<Mutex<Inner<St, F>>>),
+}
+
+impl<St, F> fmt::Debug for Partition<St, F>
+where
+    St: fmt::Debug + Stream,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Partition::Matches(inner) => f.debug_tuple("Partition::Matches").field(inner).finish(),
+            Partition::NonMatches(inner) => {
+                f.debug_tuple("Partition::NonMatches").field(inner).finish()
+            }
+        }
+    }
+}
+
+impl<St, F> Partition<St, F>
+where
+    St: Stream + Unpin,
+    F: FnMut(&St::Item) -> bool,
+{
+    pub(super) fn new(stream: St, f: F) -> (Self, Self) {
+        let inner = Arc::new(Mutex::new(Inner::new(stream, f)));
+        (
+            Partition::Matches(inner.clone()),
+            Partition::NonMatches(inner),
+        )
+    }
+}
+
+impl<St, F> Stream for Partition<St, F>
+where
+    St: Stream + Unpin,
+    F: FnMut(&St::Item) -> bool,
+{
+    type Item = St::Item;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        match self.get_mut() {
+            Partition::Matches(inner) => inner.lock().unwrap().poll_next(cx, true),
+            Partition::NonMatches(inner) => inner.lock().unwrap().poll_next(cx, false),
+        }
+    }
+}
+
+pub struct Inner<St, F>
+where
+    St: Stream,
+{
+    stream: St,
+    f: F,
+    buffered_value: Option<BufferedValue<St::Item>>,
+    waker: Option<Waker>,
+}
+
+enum BufferedValue<T> {
+    Match(T),
+    NonMatch(T),
+}
+
+impl<St, F> fmt::Debug for Inner<St, F>
+where
+    St: fmt::Debug + Stream,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Inner")
+            .field("stream", &self.stream)
+            .field("waker", &self.waker)
+            .finish_non_exhaustive()
```

Comment:
I wasn't sure about what to do about the buffered value. I don't think I can write two implementations of Debug, one where the Item doesn't implement Debug, and one where it does. I'm not sure what scenarios you'd implement Debug on a stream impl of items that aren't Debug, but I'm not sure there's a way to express that in the type system neatly.

Reply:
You can just require `St::Item: Debug`.